src/main/c/powerwaf_jni.c

```diff
@@ -899,12 +910,17 @@ static jobject _run_additive_common(JNIEnv *env, jobject this,
  * (Ljava/util/Map;Lio/sqreen/powerwaf/Powerwaf$Limits;)Lio/sqreen/powerwaf/Powerwaf$ResultWithData;
  */
 JNIEXPORT jobject JNICALL
-Java_io_sqreen_powerwaf_Additive_runAdditive__Ljava_util_Map_2Ljava_util_Map_2Lio_sqreen_powerwaf_Powerwaf_00024Limits_2Lio_sqreen_powerwaf_PowerwafMetrics_2(
-        JNIEnv *env, jobject this, jobject persistent_data, jobject ephemeral_data, jobject limits_obj,
-        jobject metrics_obj)
+Java_io_sqreen_powerwaf_Additive_runAdditive__Ljava_util_Map_2Lio_sqreen_powerwaf_Powerwaf_00024Limits_2Lio_sqreen_powerwaf_PowerwafMetrics_2Z(
+        JNIEnv *env, jobject this, jobject parameters, jobject limits_obj,
+        jobject metrics_obj, jboolean is_ephemeral)
 {
-    return _run_additive_common(env, this, persistent_data, ephemeral_data, limits_obj, metrics_obj,
-                                false);
+    if (!is_ephemeral) {
+        return _run_additive_common(env, this, parameters, NULL, limits_obj, metrics_obj,
+                                    false);
+    } else {
+        return _run_additive_common(env, this, NULL, parameters, limits_obj, metrics_obj,
+                                    false);
+    }
```

Comment:
This is not what I meant. `Additive` can perfectly take both persistent and ephemeral data at the same time.

It's `Powerwaf.runRules` (and its public caller `PowerwafContext.runRules`) where 1) the distinction ephemeral data/persistent data makes no sense and 2) where you can pass everything to libddwaf as ephemeral to get better performance.

Reply:
Fixed